pkg/trojan/message.go

```diff
@@ -0,0 +1,283 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package trojan
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+	"math/big"
+
+	"github.com/ethersphere/bee/pkg/crypto"
+	"github.com/ethersphere/bee/pkg/swarm"
+	bmtlegacy "github.com/ethersphere/bmt/legacy"
+	"golang.org/x/crypto/sha3"
+)
+
+// Topic is an alias for a 32 byte fixed-size array which contains an encoding of a message topic
+type Topic [32]byte
+
+// Target is an alias for an address which can be mined to construct a trojan message
+type Target []byte
+
+// Targets is an alias for a collection of targets
+type Targets []Target
+
+// Message represents a trojan message, which is a message that will be hidden within a chunk payload as part of its data
+type Message struct {
+	length  [2]byte // big-endian encoding of Message payload length
+	Topic   Topic
+	Payload []byte // contains the chunk address to be repaired
+	padding []byte
+}
+
+const (
+	// MaxPayloadSize is the maximum allowed payload size for the Message type, in bytes
+	MaxPayloadSize = 4030
+)
+
+var (
+	// ErrPayloadTooBig is returned when a given payload for a Message type is longer than the maximum amount allowed
+	ErrPayloadTooBig = fmt.Errorf("message payload size cannot be greater than %d bytes", MaxPayloadSize)
+
+	// ErrEmptyTargets is returned when the given target list for a trojan chunk is empty
+	ErrEmptyTargets = errors.New("target list cannot be empty")
+
+	// ErrVarLenTargets is returned when the given target list for a trojan chunk has addresses of different lengths
+	ErrVarLenTargets = errors.New("target list cannot have targets of different length")
+
+	ErrInvalidHasher = errors.New("hasher pool not initialized")
+)
+
+// NewTopic creates a new Topic variable with the given input string
+// the input string is taken as a byte slice and hashed
+func NewTopic(topic string) Topic {
+	var tpc Topic
+	t, err := crypto.LegacyKeccak256([]byte(topic))
+	if err != nil {
+		return tpc
+	}
+	copy(tpc[:], t)
+	return tpc
+}
+
+func hashFunc() hash.Hash {
+	return sha3.NewLegacyKeccak256()
+}
+
+var hasher *bmtlegacy.Hasher
+
+// NewMessage creates a new Message variable with the given topic and payload
+// it finds a length and nonce for the message according to the given input and maximum payload size
+func NewMessage(topic Topic, payload []byte) (Message, error) {
+	if len(payload) > MaxPayloadSize {
+		return Message{}, ErrPayloadTooBig
+	}
+
+	// get length as array of 2 bytes
+	payloadSize := uint16(len(payload))
+	lengthBuf := make([]byte, 2)
+	binary.BigEndian.PutUint16(lengthBuf, payloadSize)
+
+	// set random bytes as padding
+	paddingLen := MaxPayloadSize - payloadSize
```

Comment:
shouldn't there _always_ be padding? In this case, any chunk at max capacity will always be the same and have weaker security properties those who have less.

Reply:
afaik, this will always add padding as long as there is room for it

Reply:
@nolash what do you mean by "any chunk at max capacity will always be the same and have weaker security properties those who have less."

For now, the payload is the chunk address which is only 32 bytes and there will always be padding. In future if we support sending multiple chunks address that are missing .. then we might have chances that there is no padding

Reply:
> then we might have chances that there is no padding

I think we should be careful about thinking about this package only in the terms of "global pinning" (which your reply suggests). If I get my way, I will be able to craft trojan chunks independently and inject them as chunks in the system. That means that I may very well be filling data there to capacity.

usually padding is used to thwart replay-attacks in crypto. I am not entirely sure what use it has here, though?

Reply:
> usually padding is used to thwart replay-attacks in crypto. I am not entirely sure what use it has here, though?

I pass this to @zelig

Reply:
@jmozah let's let this go for now, but please make an issue out of reevaluating padding re: these questions.

Reply:
> > then we might have chances that there is no padding
> 
> I think we should be careful about thinking about this package only in the terms of "global pinning" (which your reply suggests). If I get my way, I will be able to craft trojan chunks independently and inject them as chunks in the system. That means that I may very well be filling data there to capacity more often.

this would be ideal mo